api/src/auth.rs

```diff
@@ -74,43 +79,95 @@ impl Display for AuthorizationError {
 
 impl Error for AuthorizationError {}
 
-#[rocket::async_trait]
-impl<'r> FromRequest<'r> for User {
-    type Error = AuthorizationError;
+/// A wrapper for a Rocket guard that verifies an API key is associated with a
+/// valid user.
+///
+/// The `FromRequest` impl consumes the API key and verifies it is valid for the
+/// a user. Generally you want to use [`ScopedUser`] instead to ensure the request
+/// is valid against the user's owned resources.
+#[derive(Clone, Deserialize, Serialize, Debug)]
+pub(crate) struct User {
+    pub(crate) name: String,
+    pub(crate) projects: Vec<ProjectName>,
+}
 
-    async fn from_request(req: &'r Request<'_>) -> Outcome<Self, Self::Error> {
-        let api_key = match ApiKey::try_from(req.headers().get_one("Authorization")) {
-            Ok(api_key) => api_key,
-            Err(e) => return Outcome::Failure((Status::BadRequest, e)),
-        };
-        match USER_DIRECTORY.user_for_api_key(&api_key) {
-            None => {
-                log::warn!("authorization failure for api key {:?}", &api_key);
-                Outcome::Failure((Status::Unauthorized, AuthorizationError::Unauthorized(())))
+#[async_trait]
+impl<'r> FromRequest<'r> for User {
+    type Error = Option<AuthorizationError>;
```

Comment:
Is it common to have an optional error?

Reply:
I don't think so! In this case it's not a `std::error::Error` though, it's just the associated type for [`FromRequest`](https://docs.rs/rocket/0.5.0-rc.1/rocket/request/trait.FromRequest.html). I think the idea is that's what gets printed out in server-side logs and is why this only needs to be `Debug`.

That being said, let me see if I can get rid of that `Option` 🤔